### PR TITLE
Enhance ScrollView with easing out scrolling

### PR DESCRIPTION
**Pull request by neokim**
Enhance ScrollView with elastic scroll movement.

It becomes to have iOS-like bounce-back and elastic inertia scroll using exponential ease out of tween functions.

https://www.youtube.com/watch?v=WwYJBN1bke4

It might seem that much part of sources are changed. But it is mainly resulted by code refactoring which removed some duplicated logics. No public / protected interfaces are changed and it works well with existing components.
- No attributes related to Cocos Studio are changed.
- No protected interfaces used by ListView and PageView are changed. They are not changed and work well.
- Improved maintainability. Source refactoring is done and have compact codes.
- Test project works fine.

### CODE REVIEW

**Comment by man2** [comment]
really really nice feature that cocos2d-x needed most.. thanks!!

**Comment by neokim** [comment]
Thanks for letting me know someone loves it. :)
Cocos2d-x is a great engine but I think that there are still some 'rooms' to be improved in terms of intuitive and attractive UX. I'm interested in filling those rooms.
This is a little contribution to it.

Anyway the request has been in pending status by auto build in 18 hours. Something wrong?

**Comment by zilongshanren** [comment]
@neokim  The Jenkins server hang up due to run out of free space. Now it is fixed.
Thanks for your contributing.

**Comment by zilongshanren** [comment]
[ci rebuild]

**Comment by neokim** [comment]
Jenkins failed by the following exception in Mac configuration. (http://45.56.80.45:8080/job/ccs-pr/1527/node=mac/console)

@andyque can you take a look at it?

```
FATAL: java.io.IOException: Unexpected termination of the channel
hudson.remoting.RequestAbortedException: java.io.IOException: Unexpected termination of the channel
    at hudson.remoting.Request.abort(Request.java:296)
    at hudson.remoting.Channel.terminate(Channel.java:815)
    at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:69)
    at ......remote call to mac(Native Method)
    at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1361)
    at hudson.remoting.Request.call(Request.java:171)
    at hudson.remoting.Channel.call(Channel.java:752)
    at hudson.Launcher$RemoteLauncher.kill(Launcher.java:954)
    at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:543)
    at hudson.model.Run.execute(Run.java:1744)
    at hudson.matrix.MatrixRun.run(MatrixRun.java:146)
    at hudson.model.ResourceController.execute(ResourceController.java:98)
    at hudson.model.Executor.run(Executor.java:374)
Caused by: java.io.IOException: Unexpected termination of the channel
    at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:50)
Caused by: java.io.EOFException
    at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2325)
    at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:2794)
    at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:801)
    at java.io.ObjectInputStream.<init>(ObjectInputStream.java:299)
    at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:40)
    at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
    at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:48)
Finished: FAILURE
```

**Comment by zilongshanren** [comment]
@neokim  oops, the mac slave is not stable. I think your PR is ok. 
I will review it later, thanks.

**Comment by zilongshanren** [comment]
[ci rebuild]

File: cocos/ui/UIScrollView.cpp [hunk 1]
```diff
@@ -292,932 +261,309 @@ Node* ScrollView::getChildByName(const std::string& name)const
...
-            }
-            if (des.x <= 0)
-            {
-                finalOffsetX = MAX(des.x, _contentSize.width - _innerContainer->getContentSize().width);
-            }
-            break;
-        default:
-            break;
-    }
+	if (des.y <= 0 && _direction != Direction::HORIZONTAL)
```
Comment (zilongshanren) [inline cocos/ui/UIScrollView.cpp]: nice refactor, this simplify the logic a lot.

**Comment by zilongshanren** [comment]
@neokim  A big thanks for your contribution. :+1: 
You have done the refactoring which I have dreamed to refactor for month.
Now the code is more cleaner and the bouncing logic is fancy.
It's a big win for Cocos2D-X developers. 
Thanks very much! :)

**Comment by neokim** [comment]
It was my pleasure. :)
Actually most things were already there, I just added a little feature.

**Pull request merged**